### PR TITLE
[Feat]: Add Final Norm for vLLM Hidden Extractor

### PR DESCRIPTION
**Pull request by h-guo18**
### What does this PR do?

**Type of change:** Bug fix

vLLM captures the final-layer hidden state *before* the model's final norm, but the
offline/streaming distillation path fed it straight into `lm_head`, so the reconstructed
base logits (the KD target) were computed from un-normed hidden states.

This PR re-applies the base model's final norm before `lm_head` when the producer declares
a pre-norm capture (`base_hidden_prenorm`), for both DFlash and EAGLE:

- Producer sets `base_hidden_prenorm` (streaming: `True`; offline: from the dump).
- Consumer (`_maybe_apply_base_final_norm`) re-applies the base final norm, and **fails loud**
  if pre-norm is declared but the model's norm type isn't supported (no silent corruption).
- `FakeBaseModel` now loads the base final norm (+ `rope_theta`/`rms_norm_eps`); norm type is
  gated by an explicit `model_type` allowlist (gpt_oss excluded pending a matching norm class).

### Testing

`tests/unit/torch/speculative/plugins/test_modeling_final_norm.py`; DFlash/EAGLE streaming
training verified end-to-end.

- Backward compatible?: ✅ (post-norm captures declare `base_hidden_prenorm=False` → unchanged)
- New tests: ✅

## Summary by CodeRabbit

* **New Features**
  * Expanded speculative decoding/distillation to reconstruct missing base-model logits by optionally applying a base model’s final pre–LM-head normalization when pre-norm hidden states are provided.
  * Streaming dataset now emits `base_hidden_prenorm` and can configure RDMA backends from environment settings.
* **Bug Fixes**
  * Rejects mixed `base_hidden_prenorm` values within a batch.
  * Fails fast on streaming token-length mismatches.
  * Streaming dataset loading supports directory inputs by expanding sorted JSONL shards (and errors if none are found).
* **Tests**
  * Added/updated unit and dataset tests for final-norm behavior and the new batch field.

### CODE REVIEW

File: tests/unit/torch/speculative/plugins/test_modeling_final_norm.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        # gpt_oss is intentionally DISABLED: its RMSNorm forward/weight dtype differs from the
+        # Llama variant we reuse, so it must not resolve to a norm type until a matching class
+        # is added.
+        ("gpt_oss", None),
+        ("gemma", None),  # unlisted model
+        ("", None),
+        (None, None),
+    ],
+)
+def test_select_final_norm_type(model_type, expected):
```
Comment (h-guo18) [inline tests/unit/torch/speculative/plugins/test_modeling_final_norm.py:49]: I think this test is redundant. Remove

File: modelopt/torch/speculative/plugins/rdma_hidden_states_connector.py [hunk 2]
```diff
@@ -249,7 +249,14 @@ def register_kv_caches(self, kv_caches: dict[str, torch.Tensor]):
                 max_num_seqs,
             )
 
-        self._nixl = nixl_agent(f"hs-producer-{uuid.uuid4()}", nixl_agent_config(backends=["UCX"]))
+        # Backend(s) from NIXL_BACKENDS; shared helper keeps this locked to the trainer-side
+        # agent (they must use the same backend to hand off over RDMA).
+        from .hf_streaming_dataset import nixl_backends_from_env
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/rdma_hidden_states_connector.py]: Remove the comment. Move this import to top of the file

File: modelopt/torch/speculative/plugins/modeling_final_norm.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+_FINAL_NORM_TYPE_BY_MODEL_TYPE: dict[str, str] = {
+    "llama": "rmsnorm",
+    "mistral": "rmsnorm",
+    "mixtral": "rmsnorm",
+    "qwen2": "rmsnorm",
+    "qwen3": "rmsnorm",
+    "qwen3_moe": "rmsnorm",
+    "deepseek_v3": "rmsnorm",
+    "kimi_k25": "rmsnorm",  # Kimi-K2.5 / K2.6 / K2.7 all report model_type "kimi_k25"
+    # NOTE: gpt_oss is intentionally DISABLED for now. Its GptOssRMSNorm keeps the weight in
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/modeling_final_norm.py]: trim comment. too long.

File: modelopt/torch/speculative/plugins/modeling_final_norm.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+# Registry of self-implemented final-norm variants. We deliberately reimplement these
+# (rather than importing the base model's actual module) to keep FakeBaseModel lightweight.
+# Only a small, explicit set is supported; add a class here when a new type is needed.
+_FINAL_NORM_CLASSES = {
+    "rmsnorm": _FinalRMSNorm,
+}
+
+# Base ``model_type`` → final-norm type. ONLY listed models get a final norm: we apply it
+# solely when we know which norm a model uses, because the vLLM-captured final hidden is
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/modeling_final_norm.py]: make the comment more concise.

File: modelopt/torch/speculative/plugins/hf_dflash.py [hunk 5]
```diff
@@ -571,7 +597,15 @@ def forward(
...
-                out_hiddens = kwargs["base_model_outputs"]["base_model_hidden_states"]
+                bmo = kwargs["base_model_outputs"]
+                out_hiddens = bmo["base_model_hidden_states"]
+                # Re-apply the base model's final norm before lm_head when the producer captured
+                # a pre-(final-)norm hidden (vLLM does; HF/TRT-LLM capture post-norm and declare
+                # base_hidden_prenorm False). Honoring the producer's declaration keeps this
+                # consumer backend-agnostic. If the producer says pre-norm but no final norm was
+                # located, we CANNOT reconstruct correct logits — fail loud rather than silently
+                # feeding an un-normed hidden into lm_head (a corrupt distillation target).
+                out_hiddens = _maybe_apply_base_final_norm(out_hiddens, bmo, self._base_model_norm)
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/hf_dflash.py]: Can we move this logic of optionally applying final norm to DFlashBaseModelOutput, such that, DFlashBaseModelOutput return a clean output hiddens regardless format of from_offline_dict

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py [hunk 6]
```diff
@@ -242,12 +255,16 @@ def __getitem__(self, idx: int) -> dict[str, torch.Tensor]:
...
         """Tokenize a single entry.
 
-        Returns ``None`` for entries missing ``cid`` / ``messages``, or when
+        Returns ``None`` for entries missing ``cid`` / ``conversations``, or when
         right-truncation to ``max_seq_len`` drops the entire supervised span
         (``answer_only_loss`` mode with the assistant turn at the tail).
         """
         cid = entry.get("conversation_id") or entry.get("uuid")
-        convs = entry.get("messages") or entry.get("conversations")
+        # Only ``conversations`` is accepted. Some corpora (e.g. Spec-Decoding-Dataset-v2)
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/hf_streaming_dataset.py]: remove this comment

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py [hunk 7]
```diff
@@ -532,4 +561,7 @@ def _format(self, fetched: EagleFetchPayload) -> dict[str, torch.Tensor]:
             "attention_mask": torch.ones_like(input_ids),
             "loss_mask": loss_mask,
             "labels": labels,
+            # vLLM captures the residual stream BEFORE the final norm; tell the trainer to
+            # re-apply it before lm_head (see HFDFlashModel.forward).
+            "base_hidden_prenorm": True,
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/hf_streaming_dataset.py]: lets move this argument into self.config. It's default value True should be decalered in class EagleVllmStreamingConfig

**Review by ChenhanYu — APPROVED** [review]
